src/rust/engine/fs/src/store.rs

```diff
@@ -429,31 +433,198 @@ mod local {
     }
+  }
+}
+
+mod remote {
+  use super::Digest;
+
+  use bazel_protos;
+  use boxfuture::{Boxable, BoxFuture};
+  use futures::{future, Future, Stream};
+  use grpcio;
+  use std::sync::Arc;
+
+  use hash::Fingerprint;
+
+  #[derive(Clone)]
+  pub struct ByteStore {
+    address: String,
+    env: Arc<grpcio::Environment>,
+  }
 
-    fn load_file_bytes(
-      store: &ByteStore,
+  impl ByteStore {
+    pub fn new(address: String) -> ByteStore {
+      // TODO: Pass through a parallelism configuration option here.
+      ByteStore {
+        address: address,
+        env: Arc::new(grpcio::Environment::new(1)),
+      }
+    }
+
+    fn load_bytes_with<T: Send + 'static, F: Fn(&[u8]) -> T + Send + Sync + 'static>(
+      &self,
       fingerprint: Fingerprint,
-    ) -> Result<Option<Vec<u8>>, String> {
-      store
-        .load_file_bytes_with(fingerprint, Arc::new(|bytes: &[u8]| bytes.to_vec()))
-        .wait()
+      f: Arc<F>,
+    ) -> BoxFuture<Option<T>, String> {
+      let channel = grpcio::ChannelBuilder::new(self.env.clone()).connect(&self.address);
+      let client = bazel_protos::bytestream_grpc::ByteStreamClient::new(channel);
+      let stream = client.read(&{
+        let mut req = bazel_protos::bytestream::ReadRequest::new();
+        // TODO: Pass a size around, or resolve that we don't need to.
+        req.set_resource_name(format!("/blobs/{}/{}", fingerprint, -1));
+        req.set_read_offset(0);
+        // 0 means no limit.
+        req.set_read_limit(0);
+        req
+      });
+
+      // We shouldn't have to pass around the client here, it's a workaround for
+      // https://github.com/pingcap/grpc-rs/issues/123
+      future::ok(client)
+        .join(stream.map(|r| r.data).concat2())
+        .map(|(_client, bytes)| Some(bytes))
+        .or_else(|e| match e {
+          grpcio::Error::RpcFailure(grpcio::RpcStatus {
+                                      status: grpcio::RpcStatusCode::NotFound, ..
+                                    }) => Ok(None),
+          _ => Err(format!("Error making CAS read request: {:?}", e)),
+        })
+        .map(move |maybe_bytes| {
+          maybe_bytes.map(|bytes: Vec<u8>| f(&bytes))
+        })
+        .to_boxed()
+    }
+  }
+
+
+  impl super::ByteStore for ByteStore {
+    fn store_file_bytes(&self, _bytes: Vec<u8>) -> BoxFuture<Digest, String> {
+      unimplemented!()
     }
 
-    fn load_directory_proto_bytes(
-      store: &ByteStore,
+    fn load_file_bytes_with<T: Send + 'static, F: Fn(&[u8]) -> T + Send + Sync + 'static>(
+      &self,
       fingerprint: Fingerprint,
-    ) -> Result<Option<Vec<u8>>, String> {
-      store
-        .load_directory_proto_bytes_with(fingerprint, Arc::new(|bytes: &[u8]| bytes.to_vec()))
-        .wait()
+      f: Arc<F>,
+    ) -> BoxFuture<Option<T>, String> {
+      self.load_bytes_with(fingerprint, f)
+    }
+
+    fn record_directory(
```

Comment:
Seems like the high-level methods like `record_directory` and `load_directory_proto_bytes_with` should move onto the top-level `impl Store`, and the implementations of `ByteStore` should just deal with bytes? Alternatively, they could become concrete methods of `ByteStore` trait.

One thing that might help with that is: rather than passing around `Database` handles, you could add:
```
enum EntryType {
  Directory,
  File
}
```

... and add it to the signature of `store_bytes`/`load_bytes*`.

Reply:
Done